src/backend/gl/src/device.rs

```diff
@@ -464,73 +463,81 @@ impl d::Device<B> for Device {
+                        0
+                    } else {
+                        assert!(is_cpu_visible_memory);
+                        let read_flag = if is_readable_memory {
+                            glow::MAP_READ_BIT
+                        } else {
+                            0
+                        };
+                        read_flag
+                            | glow::MAP_WRITE_BIT
+                            | glow::MAP_PERSISTENT_BIT
+                            | glow::DYNAMIC_STORAGE_BIT
+                    };
+                    gl.buffer_storage(target, size as i32, None, flags);
                 } else {
-                    glow::DYNAMIC_DRAW
-                };
-                gl.buffer_data_size(target, size as i32, usage);
-            }
-            gl.bind_buffer(target, None);
+                    let usage = if is_device_local_memory {
+                        glow::STATIC_DRAW
+                    } else {
+                        glow::DYNAMIC_DRAW
+                    };
+                    gl.buffer_data_size(target, size as i32, usage);
+                }
+                gl.bind_buffer(target, None);
 
-            if let Err(err) = self.share.check() {
-                panic!("Error allocating memory buffer {:?}", err);
-            }
+                if let Err(err) = self.share.check() {
+                    panic!("Error allocating memory buffer {:?}", err);
+                }
 
-            let mut map_flags = glow::MAP_READ_BIT | glow::MAP_WRITE_BIT | glow::MAP_PERSISTENT_BIT;
-            if !is_coherent_memory {
-                map_flags |= glow::MAP_FLUSH_EXPLICIT_BIT;
+                let mut map_flags = glow::MAP_WRITE_BIT
+                    | glow::MAP_PERSISTENT_BIT
+                    | glow::MAP_FLUSH_EXPLICIT_BIT;
+                if is_readable_memory {
+                    map_flags |= glow::MAP_READ_BIT;
+                }
+                if !is_coherent_memory {
+                    map_flags |= glow::MAP_COHERENT_BIT;
+                }
+
+                Ok(n::Memory {
+                    properties: memory_type.properties,
+                    buffer: Some(raw),
+                    size,
+                    target,
+                    map_flags,
+                    emulate_map_allocation: Cell::new(None),
+                })
             }
 
-            Ok(n::Memory {
-                properties: memory::Properties::CPU_VISIBLE
-                    | memory::Properties::CPU_CACHED,
-                buffer: Some(raw),
-                target,
-                size,
-                emulate_map_allocation: Cell::new(None),
-                map_flags,
-            })
+            MemoryRole::Image => {
+                assert!(is_device_local_memory);
+                Ok(n::Memory {
+                    properties: memory::Properties::DEVICE_LOCAL,
+                    buffer: None,
+                    size,
+                    target: 0,
```

Comment:
let's move this field behind the `buffer` option?

Reply:
Yeah, I almost did that before actually...